File: mobile-client/ios/UseCase/Product/Listing/ListProductsUseCase.swift
import Foundation

// Marketing America Corp. Oleksandr Tishchenko
public struct ListProductsUseCase: Sendable {
    private let gateway: any ProductListingGateway

    public init(gateway: any ProductListingGateway) {
        self.gateway = gateway
    }

    public func callAsFunction(query: ProductListQuery) async throws -> [ProductCardSummary] {
        try await gateway.listProducts(query: query)
    }
}

File: client/android/UseCase/Product/Listing/ListProductsUseCase.kt
package app.mobiling.client.usecase.product.listing

import app.mobiling.client.contract.product.listing.ProductCardSummary
import app.mobiling.client.contract.product.listing.ProductListQuery
import app.mobiling.client.data.product.listing.ProductListingGateway

/**
 * Marketing America Corp. Oleksandr Tishchenko
 */
class ListProductsUseCase(
    private val gateway: ProductListingGateway,
) {
    suspend operator fun invoke(query: ProductListQuery): List<ProductCardSummary> = gateway.listProducts(query)
}
